### PR TITLE
Project Exporter Refactor

### PR DESCRIPTION
**Pull request by spham-amzn**
## What does this PR do?

This is a refactor of project export which does the following:
* Combined engine-centric and project-centric workflows
* Added an extra '/' separator arg in order to pass through additional cmake build args
* Added Unit Tests

Fixes #16562 

## How was this PR tested?

* Unit Tested
```
> python\python.cmd -m pytest scripts\o3de
Pytest custom argument "--build-directory" was not provided, tests using LyTestTools will fail
Pytest custom argument "--output-path" was not provided, defaulting Test Results output to: D:\github\o3de\pytest_results\2023-08-15T11-37-47-623765
================================================================================ test session starts ================================================================================= platform win32 -- Python 3.10.5, pytest-7.2.0, pluggy-0.13.1
rootdir: D:\github\o3de, configfile: pytest.ini
plugins: mock-3.8.2, timeout-2.1.0, ly-test-tools-1.0.0
timeout: 1200.0s
timeout method: thread
timeout func_only: False
collected 435 items

scripts\o3de\tests\test_cmake.py ...................                                                                                                                            [  4%]
scripts\o3de\tests\test_compatibility.py ...........................                                                                                                            [ 10%] 
scripts\o3de\tests\test_disable_gem.py .........                                                                                                                                [ 12%] 
scripts\o3de\tests\test_download.py .....................                                                                                                                       [ 17%]
scripts\o3de\tests\test_enable_gem.py ....................................                                                                                                      [ 25%] scripts\o3de\tests\test_engine_properties.py ....                                                                                                                               [ 26%]
scripts\o3de\tests\test_engine_template.py ............                                                                                                                         [ 29%]
scripts\o3de\tests\test_export_project.py ............................                                                                                                          [ 35%] scripts\o3de\tests\test_gem_properties.py .....                                                                                                                                 [ 37%]
scripts\o3de\tests\test_global_project.py ..                                                                                                                                    [ 37%]
scripts\o3de\tests\test_manifest.py ....................................................                                                                                        [ 49%]
scripts\o3de\tests\test_print_registration.py ..............................................                                                                                    [ 60%]
 scripts\o3de\tests\test_project_manager_interface.py ....................................                                                                                       [ 68%] 
scripts\o3de\tests\test_project_properties.py .........                                                                                                                         [ 70%] 
scripts\o3de\tests\test_register.py ....................................                                                                                                        [ 78%] 
scripts\o3de\tests\test_repo.py .................                                                                                                                               [ 82%] 
scripts\o3de\tests\test_repo_properties.py ............................                                                                                                         [ 88%]
scripts\o3de\tests\test_utils.py ................................................                                                                                               [100%]

================================================================================ 435 passed in 6.73s =================================================================================
```
* Run project-centric with the following command:
```
λ %O3DE_PATH%\scripts\o3de export-project -es %O3DE_PATH%\scripts\o3de\ExportScripts\export_source_built_project.py -pp %O3DE_PROJECT_PATH% -out %OUTPUT_PATH% -cfg release -a zip -noified -gpfp launch_client.cfg -spfp launch_client.cfg -code -assets -ll INFO -sl %O3DE_PROJECT_PATH%\AssetBundling\SeedLists\BasePopcornFxSeedList.seed -sl %O3DE_PROJECT_PATH%\AssetBundling\SeedLists\GameSeedList.seed -sl %O3DE_PROJECT_PATH%\AssetBundling\SeedLists\VFXSeedList.seed ```

### CODE REVIEW

File: scripts/o3de/ExportScripts/export_source_built_project.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+# This script is meant to be run via o3de cli export-project command, not from Python directly!
+# This export script supports optional building of asset tools, so it is only compatible with the source distribution
+# of O3DE (https://github.com/o3de/o3de)
+
+# Example invocation on Linux
+'''
+$ cd <project-path>
+$ <engine-path>/scripts/o3de.sh export-project --export-scripts ExportScripts/export_source_built_project.py \
+ --project-path . --log-level INFO --output-path <install-path> --build-non-mono-tools --config release --archive-output zip \
+ --seedlist ./Assets/seedlist1.seed --seedlist ./Assets/seedlist2.seed <etc..>
```
Comment (AMZN-Gene) [inline scripts/o3de/ExportScripts/export_source_built_project.py:33]:
```suggestion
 --seedlist ./Assets/seedlist1.seed --seedlist ./Assets/seedlist2.seed <etc..> --should-build-code --should-build-assets
```

File: scripts/o3de/ExportScripts/export_source_built_project.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+ --project-path . --log-level INFO --output-path <install-path> --build-non-mono-tools --config release --archive-output zip \
+ --seedlist ./Assets/seedlist1.seed --seedlist ./Assets/seedlist2.seed <etc..> 
+'''
+
+# Example invocation on Windows
+'''
+$ cd <project-path>
+$ <engine-path>\scripts\o3de.bat  export-project --export-scripts ExportScripts\eexport_source_built_project.py \
+ --project-path . --log-level INFO --output-path <install-path> --build-non-mono-tools --config release --archive-output zip \
+ --seedlist .\Assets\seedlist1.seed --seedlist .\Assets\seedlist2.seed <etc..>
```
Comment (AMZN-Gene) [inline scripts/o3de/ExportScripts/export_source_built_project.py:41]:
```suggestion
--seedlist .\Assets\seedlist1.seed --seedlist .\Assets\seedlist2.seed <etc..> --should-build-code --should-build-assets
```

File: scripts/o3de/o3de/export_project.py [hunk 3]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
...
+
+    return bundles_path
+
+
+def kill_existing_processes(project_name: str):
+    o3de_process_names = ['o3de', 'editor', 'assetprocessor', f'{project_name.lower()}.serverlauncher',
+                          f'{project_name.lower()}.gamelauncher']
+    processes_to_kill = []
+    for process in psutil.process_iter():
+
```
Comment (AMZN-Gene) [inline scripts/o3de/o3de/export_project.py:454]:
```suggestion
    # Don't try to close o3de.sh (the script currently running), but still look for o3de (Project Manager)
    if process.name() == 'o3de.sh'
        continue
```

**Review by AMZN-Gene — APPROVED** [review]

File: scripts/o3de/o3de/export_project.py (hunk 3 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:454]: That check is already handled in the line below with the `O3DE_SCRIPT_NAME` global variable

**Review by tkothadev — CHANGES_REQUESTED** [review]
Could you make sure to also update the export scripts found in the project template folders for `DefaultProject` and `MinimalProject`. I think they use the outdated script atm. That will also need to be tested to see if a) the updated export script is correctly generated from template, and b) that script correctly runs

File: scripts/o3de/o3de/export_project.py [hunk 4]
```diff
@@ -7,28 +7,69 @@
...
+# Account for some windows-specific attributes
+if platform.system().lower() == 'windows':
+    EXECUTABLE_EXTENSION = '.exe'
+    O3DE_SCRIPT_NAME = 'o3de.bat'
+    GENERATOR = None
+    CMAKE_GENERATOR_OPTIONS = ['-DLY_DISABLE_TEST_MODULES=ON']
+else:
+    EXECUTABLE_EXTENSION = ""
+    O3DE_SCRIPT_NAME = 'o3de.sh'
+    GENERATOR = "Ninja Multi-Config"
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:31]: Should we always auto assign Ninja generator for unix? There may be use cases where Unix Makefiles are preferred (is there an option to set/override the default generator?)

File: scripts/o3de/o3de/export_project.py [hunk 5]
```diff
@@ -7,28 +7,69 @@
...
         self._args = args
+        self._cmake_additional_build_args = cmake_additional_build_args
+
+        project_json_data = manifest.get_project_json_data(project_path=project_path)
+        assert project_json_data, f"Invalid project configuration file '{project_path}/project.json'. Invalid settings."
+
+        project_name = project_json_data.get('project_name')
+        assert project_name, f"Invalid project configuration file '{project_path}/project.json'. 'project_name' not found in the settings"
+
+        self._project_name = project_name
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:70]: would it make sense to just include the project json data overall?

File: scripts/o3de/o3de/export_project.py [hunk 6]
```diff
@@ -95,19 +152,33 @@ def _export_script(export_script_path: pathlib.Path, project_path: pathlib.Path,
...
             logging.error(f"Unable to find project folder associated with file '{export_script_path}'. Please specify using --project-path, or ensure the file is inside a project folder.")
         return 1
-    
-    o3de_context = O3DEScriptExportContext(export_script_path= export_script_path,
-                                        project_path = computed_project_path,
-                                        engine_path = manifest.get_project_engine_path(computed_project_path),
-                                        args = passthru_args)
+
+    # Collect and isolate any custom cmake build arguments that may have been passed through, use `/` as the separator in the args
+    export_process_args = []
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:165]: If I understand this code segment correctly, you're trying to split between cmake specific arguments and exporter arguments, something like:

`export arg 1  arg 2 .. / cmake arg 1 arg 2 ..`

in that case, you may want to first find the index of the separator, if it exists, and then slice the list you want:

```
separator_string = '/'

sep_index= passthru_args.index(separator_string)
cmake_custom_build_args = passthru_args[sep_index + 1:] #slice will select relevant cmake args after separator
export_process_args = passthru_args[:sep_index] #omit the separator
```

File: scripts/o3de/o3de/export_project.py [hunk 7]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
     export_subparser = subparsers.add_parser('export-project')
     add_parser_args(export_subparser)
 
+
+def get_asset_processor_batch_path(non_mono_build_path: pathlib.Path):
+    return non_mono_build_path / f'bin/profile/AssetProcessorBatch{EXECUTABLE_EXTENSION}'
+
+
+def get_asset_bundler_batch_path(non_mono_build_path: pathlib.Path):
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:215]: suggestion: rename `non_mono_build_path` to `tool_root_path`

File: scripts/o3de/o3de/export_project.py [hunk 8]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
     export_subparser = subparsers.add_parser('export-project')
     add_parser_args(export_subparser)
 
+
+def get_asset_processor_batch_path(non_mono_build_path: pathlib.Path):
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:211]: suggestion: rename `non_mono_build_path` to `tool_root_path`

File: scripts/o3de/ExportScripts/export_source_built_project.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+import shutil
+
+from o3de.export_project import process_command
+from o3de import manifest
+from typing import List
+
+# This script is meant to be run via o3de cli export-project command, not from Python directly!
+# This export script supports optional building of asset tools, so it is only compatible with the source distribution
+# of O3DE (https://github.com/o3de/o3de)
+
```
Comment (tkothadev) [inline scripts/o3de/ExportScripts/export_source_built_project.py:27]: should we update these instructions to include specific examples for passing in custom CMake args?

File: scripts/o3de/o3de/export_project.py (hunk 4 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:31]: In the android setup scripts, it tries to auto-detect if ninja is available before using it or falling back to unix make files. I can add that kind of logic here as well

File: scripts/o3de/ExportScripts/export_source_built_project.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+        expected_bundles_path = exp.bundle_assets(ctx=ctx,
+                                                  selected_platform=selected_platform,
+                                                  seedlist_paths=seedlist_paths,
+                                                  non_mono_build_path=tools_build_path,
+                                                  custom_asset_list_path=preferred_asset_list_path,
+                                                  max_bundle_size=max_bundle_size)
+    else:
+        expected_bundles_path = ctx.project_path / "AssetBundling/Bundles"
+
+    for should_do, expected_output_dir, project_file_patterns, ignore_patterns in zip(
```
Comment (tkothadev) [inline scripts/o3de/ExportScripts/export_source_built_project.py:126]: I'm beginning to think the zipped objects warrant a dedicated class structure for the project export script. Something like this class definition:

```python
class SetupLayoutConfig:
   def __init__(self, output_path, should_build = True, project_file_patterns = [], ignore_patterns = [] ):
      self.should_build = should_build
      self.output_path = output_path
      self.project_file_patterns = project_file_patterns
      self.ignore_patterns = ignore_patterns
```

File: scripts/o3de/o3de/export_project.py (hunk 6 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:165]: thats a good idea

File: scripts/o3de/ExportScripts/export_source_built_project.py (hunk 10 above)
Comment (tkothadev) [reply scripts/o3de/ExportScripts/export_source_built_project.py:126]: It would make it easier to discern which settings belong to which use case

Comment (tkothadev) [reply scripts/o3de/ExportScripts/export_source_built_project.py:126]: then just iterate over the list of specified launchers

File: scripts/o3de/o3de/export_project.py [hunk 11]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
...
+    cmake_configure_command = ["cmake", "-B", tools_build_path]
+
+    if ctx.is_engine_centric:
+        cmake_configure_command.extend(["-S", ctx.engine_path])
+    else:
+        cmake_configure_command.extend(["-S", ctx.project_path])
+
+    # Generate the project for the pre-requisite tools
+    if GENERATOR:
+        cmake_configure_command.extend(["-G", GENERATOR])
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:277]: If Generator is always auto-assigned at the start of `export_project.py`, then is there anyway to override the preferred generator if needed?

File: scripts/o3de/o3de/export_project.py [hunk 12]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
...
+
+    # Build the project for the pre-requisite tools
+    cmake_build_command = ["cmake", "--build", tools_build_path,
+                                    "--config", PREREQUISITE_TOOL_BUILD_CONFIG,
+                                    "--target", "AssetProcessorBatch", "AssetBundlerBatch"]
+    if ctx.cmake_additional_build_args:
+        cmake_build_command.extend(ctx.cmake_additional_build_args)
+    if logger:
+        logger.info(f"Building tool chain files for project {ctx.project_name}.")
+    ret = process_command(cmake_build_command, )
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:296]: nit: trailing comma

File: scripts/o3de/o3de/export_project.py [hunk 13]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
...
+    @param ctx:                         Export Context
+    @param selected_platform:           The desired asset platform
+    @param seedlist_paths:              The list of seedlist files
+    @param non_mono_build_path:         The path to the tools cmake build project
+    @param custom_asset_list_path:      Optional custom asset list path, otherwise use the expectged 'AssetBundling/AssetLists'path in the project folder
+    @param max_bundle_size:             The size limit to put on the bundle
+    @return: The path to the bundle
+    """
+
+    asset_bundler_batch_path = get_asset_bundler_batch_path(non_mono_build_path)
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:389]: Could we add a check for verifying the asset bundler's existence?

File: scripts/o3de/o3de/export_project.py [hunk 14]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
...
+    ret = process_command(mono_build_args)
+    if ret != 0:
+        raise ExportProjectError("Error building the monolithic launcher(s).")
+    if logger:
+        logger.info(f"Project {ctx.project_name} built successfully.")
+
+
+def bundle_assets(ctx: O3DEScriptExportContext,
+                  selected_platform: str,
+                  seedlist_paths: List[str],
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:373]: should this be a `List[pathlib.Path]` instead?

File: scripts/o3de/o3de/export_project.py [hunk 15]
```diff
@@ -136,3 +207,322 @@ def add_args(subparsers) -> None:
...
+    output_cache_path = output_path / 'Cache' / asset_platform
+
+    os.makedirs(output_cache_path, exist_ok=True)
+
+    for bundle in bundles_to_copy:
+        shutil.copy(bundle, output_cache_path)
+
+    for file in glob.glob(str(mono_build_path / f'bin/{build_config}/*')):
+        file_path = pathlib.Path(file)
+        if file_path.is_dir():
```
Comment (tkothadev) [inline scripts/o3de/o3de/export_project.py:508]: do we need to extend ignore patterns to also consider folders too?

File: scripts/o3de/ExportScripts/export_source_built_project.py (hunk 10 above)
Comment (spham-amzn) [reply scripts/o3de/ExportScripts/export_source_built_project.py:126]: +1 Most definitely

File: scripts/o3de/o3de/export_project.py (hunk 11 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:277]: I think we should keep the options as light as possible and not introduce any options, especially since the generator doesn't provide us any benefits for this process other than Ninja supporting parallel builds (Unix doesnt), but that will be addressed as mentioned above

Comment (tkothadev) [reply scripts/o3de/o3de/export_project.py:277]: that's fair, I think we will only introduce it if there's a real need for it

File: scripts/o3de/o3de/export_project.py (hunk 5 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:70]: Not really, unless there is a need. Its better to keep the pattern to the minimal information that is needed

**Comment by tkothadev** [comment]
I do remember we discussed removing the project export script from the template folders, but could we add a readme file as a stub containing instructions for where to find reference scripts?

File: scripts/o3de/ExportScripts/export_source_built_project.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+import logging
+import sys
+
+import o3de.export_project as exp
+import pathlib
+
+from typing import List
+
+
+def export_standalone_monolithic_project_centric(ctx: exp.O3DEScriptExportContext,
```
Comment (tkothadev) [inline scripts/o3de/ExportScripts/export_source_built_project.py:19]: There are some parts of the function logic which checks for engine_centric build in the context. Maybe we should strip `project_centric` from the function name?

File: scripts/o3de/o3de/utils.py [hunk 17]
```diff
@@ -135,25 +134,20 @@ def run(self) -> int:
         """
         ret = 1
         try:
-            with Popen(self.args, cwd=self.cwd, env=self.env, stdout=PIPE, stderr=PIPE) as process:
+            with Popen(self.args, cwd=self.cwd, env=self.env, stdout=PIPE, stderr=STDOUT) as process:
```
Comment (AMZN-Gene) [inline scripts/o3de/o3de/utils.py:137]: Could you update the PR comment to mention why the change to STDOUT?

**Review by AMZN-Gene — APPROVED** [review]

File: scripts/o3de/o3de/utils.py (hunk 17 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/utils.py:137]: setting `stderr` to `PIPE` forces us to read the error stream at the end of the process. During the polling process, you cannot poll `stderr` because it was blocking. `STDOUT` will redirect to `stdout` which is directed to `PIPE`, so instead we are reporting the error lines as they occur in sequence, not the end

**Comment by spham-amzn** [comment]
> I do remember we discussed removing the project export script from the template folders, but could we add a readme file in the ExportScripts folders of the templates as a stub containing instructions for where to find reference scripts?

There is a README.md which describes the `ExportScripts` files. I'll add information in there instead of how to customize a project export to suit their specific project. I don't think its necessary to do so in the project template, we want to minimize setting anything that can be changed in the engine (like bug fixes) so that errors or changes won't live in the generated projects forever

File: scripts/o3de/ExportScripts/README.md [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+**Linux**
+```
+EXPORT O3DE_PATH=$HOME/o3de
+EXPORT O3DE_PROJECT_PATH=$HOME/O3DE/NewProject
+EXPORT O3DE_PROJECT_SEEDLIST=$O3DE_PROJECT_PATH/Assets/seedlist.seed
+EXPORT OUTPUT_PATH=$HOME/NewProjectPackages
+
+$O3DE_PATH/scripts/o3de.sh export-project --export-scripts ExportScripts/export_source_built_project.py 
+ --project-path $O3DE_PROJECT_PATH --log-level INFO --output-path %OUTPUT_PATH% --build-tools --config release --archive-output zip \
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/README.md:24]: The linux instructions should use `$` not `%`
```suggestion
 --project-path $O3DE_PROJECT_PATH --log-level INFO --output-path $OUTPUT_PATH --build-tools --config release --archive-output zip \
```

File: scripts/o3de/ExportScripts/export_source_built_project.py [hunk 19]
```diff
@@ -0,0 +1,252 @@
...
+    if not logger:
+        logger = logging.getLogger()
+        logger.setLevel(logging.ERROR)
+
+    # Calculate the tools and game build paths
+    default_base_build_path = ctx.engine_path / 'build' if ctx.is_engine_centric else ctx.project_path / 'build'
+    if not tools_build_path:
+        tools_build_path = default_base_build_path / 'tools'
+    if not game_build_path:
+        game_build_path = default_base_build_path / 'game'
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/ExportScripts/export_source_built_project.py:78]: I would suggest using the folder and variable name of 'launcher' instead of 'game'.

Using 'game' here means to configure a CMake project folder that can build ALL launchers.
While the other use of  "game" in this function such as "should_build_game_launcher" argument refers specifically to the GameLauncher.

File: scripts/o3de/o3de/export_project.py [hunk 20]
```diff
@@ -7,28 +7,94 @@
...
+        test_ninja_result = subprocess.run(['ninja', '--version'])
+    except FileNotFoundError:
+        pass
+
+    if test_ninja_result and test_ninja_result.returncode == 0:
+        GENERATOR = "Ninja Multi-Config"
+        CMAKE_MULTI_CONFIGURATION_GENERATOR = True
+    else:
+        GENERATOR = "Unix Makefiles"
+        CMAKE_MULTI_CONFIGURATION_GENERATOR = False
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/export_project.py:48]: This will also need to check for Xcode if running on MacOS

File: scripts/o3de/o3de/export_project.py [hunk 21]
```diff
@@ -7,28 +7,94 @@
...
+        self._cmake_additional_build_args = cmake_additional_build_args
+
+        project_json_data = manifest.get_project_json_data(project_path=project_path)
+        assert project_json_data, f"Invalid project configuration file '{project_path}/project.json'. Invalid settings."
+
+        project_name = project_json_data.get('project_name')
+        assert project_name, f"Invalid project configuration file '{project_path}/project.json'. 'project_name' not found in the settings"
+
+        self._project_name = project_name
+        self._is_engine_centric = project_path.is_relative_to(engine_path)
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/export_project.py:96]: What it means to be engine centric vs project-centric is what is used as the root CMakeLists.txt for cmake configuration?

For example if you configured with the AutomatedTesting directory as the source directory for cmake, then that is project-centric, since it uses the project CMakeLists.txt for as the cmake entry point.

```bash
[spham-amzn@ubuntu ~/o3de]$ cmake -B AutomatedTesting/build/linux -S AutomatedTesting
```

If you use the engine directory as the cmake source directory, then that is engine-centric.
The initial cmake entry in that case is the CMakeLists.txt in the root of the engine
```bash
[spham-amzn@ubuntu ~/o3de]$ cmake -B AutomatedTesting/build/linux -S . -DLY_PROJECT=./AutomatedTesting
```

What this means is that unless you ask the user for whether they want to use the project root or engine root as the Top level cmake source directory, then the code is just guessing at which one to use.

File: scripts/o3de/o3de/export_project.py [hunk 22]
```diff
@@ -136,3 +283,402 @@ def add_args(subparsers) -> None:
...
     add_parser_args(export_subparser)
 
+
+def get_asset_processor_batch_path(tools_build_path: pathlib.Path,
+                                   required: bool = False) -> pathlib.Path:
+    """
+    Get the expected path to the asset bundler tool
+
+    @param tools_build_path:    The tools (cmake) build path to locate AssetProcessorBatch
+    @param required:            If true, check if the asset bundler actually exists on file at the expected location, and raise an error if not
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/export_project.py:293]: The comment here should mention the asset processor and not bundler
```suggestion
    @param required:            If true, check if the asset processor actually exists on file at the expected location, and raise an error if not
```

File: scripts/o3de/o3de/export_project.py [hunk 23]
```diff
@@ -136,3 +283,402 @@ def add_args(subparsers) -> None:
...
+    ret = process_command(cmake_build_assets_command,
+                          cwd=ctx.engine_path if ctx.is_engine_centric else ctx.project_path)
+    if ret != 0:
+        if fail_on_ap_errors:
+            raise ExportProjectError(f"Error building assets for project {ctx.project_name}.")
+        else:
+            if logger:
+                logger.warning("Some assets failed to processed.")
+    if logger:
+        logger.info(f"Completed processing assets for {ctx.project_name}!")
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/export_project.py:350]: This seems like it should return the AssetProcessorBatch return code at the end of this function when `fail_on_ap_errors` is false.

File: scripts/o3de/o3de/export_project.py [hunk 24]
```diff
@@ -136,3 +283,402 @@ def add_args(subparsers) -> None:
...
+    if ctx.is_engine_centric:
+        cmake_configure_command.extend(["-S", ctx.engine_path])
+    else:
+        cmake_configure_command.extend(["-S", ctx.project_path])
+
+    # Generate the project for the pre-requisite tools
+    if GENERATOR:
+        cmake_configure_command.extend(["-G", GENERATOR])
+    if CMAKE_GENERATOR_OPTIONS:
+        cmake_configure_command.extend(CMAKE_GENERATOR_OPTIONS)
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/export_project.py:377]: We don't have an argparse option for specifying an additional cmake configure argument (such as a `--preset` or any custom options the user want to configure with)

File: scripts/o3de/o3de/export_project.py [hunk 25]
```diff
@@ -136,3 +283,402 @@ def add_args(subparsers) -> None:
...
+
+    @param tools_build_path:    The base tools build path
+    """
+    # Otherwise make sure the tools exist already
+    build_tools = [get_asset_bundler_batch_path(tools_build_path),
+                   get_asset_processor_batch_path(tools_build_path)]
+    tools_missing = [b for b in build_tools if not b.exists()]
+    if len(tools_missing) > 0:
+        raise ExportProjectError(f"Necessary Build Tools have not been created! The following are missing: {', '.join(tools_missing)}"
+                                 "Please ensure that these tools exist before proceeding on with the build!")
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/export_project.py:419]: I would only throw an exception if these tools need to be run.
If user has already built and bundled their assets and only need to re-build the launchers, then this doesn't need to error.

File: scripts/o3de/o3de/export_project.py (hunk 21 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:96]: If that is the case, and its possible to set `-DLY_PROJECT` to an absolute path, then its probably better to make an engine build a switch (argument) that is passed in instead of assuming a project folder outside of the engine root is project centric, etc.

Comment (lumberyard-employee-dm) [reply scripts/o3de/o3de/export_project.py:96]: Yes, `LY_PROJECT` accepts a path that can be absolute or relative to the [cmake source directory](https://github.com/o3de/o3de/blob/development/cmake/Projects.cmake#L263).
With the information that the script has, it can always choose either project-centric or engine-centric in every situation.

File: scripts/o3de/o3de/export_project.py (hunk 22 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:293]: Was also wrong in the description

File: scripts/o3de/o3de/export_project.py (hunk 24 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:377]: I don't want to introduce more complexity to the arguments at this time. I only added the ability to pass through the cmake build commands because it affects performance of the build.

File: scripts/o3de/o3de/export_project.py (hunk 25 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:419]: Right now the there is a 'should_build_all_assets' which should probably need to change to 'should_build_asset_and_bundle' or 'should_build_bundle' since building the assets should be implied if you want to rebuild the bundle. I would have to make this change and rely on the flag to check on the tools

Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:419]: There was no real need for this function, since we want a more granular check, i will check each tool if needed

File: scripts/o3de/o3de/export_project.py (hunk 24 above)
Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:401]: I just realized that I may need to handle this situation. For simulation build workflows, PhysX 5 is needed, but that is an opt-in CMAKE configure parameter that needs to be set.

File: scripts/o3de/o3de/export_project.py [hunk 26]
```diff
@@ -403,6 +406,8 @@ def build_export_toolchain(ctx: O3DEScriptExportContext,
         cmake_configure_command.extend([f'-DCMAKE_BUILD_TYPE={PREREQUISITE_TOOL_BUILD_CONFIG}'])
     if engine_centric:
         cmake_configure_command.extend([f'-DLY_PROJECTS={ctx.project_path}'])
+    if additional_cmake_configure_options:
+        cmake_configure_command.extend([f'-D{option}' for option in additional_cmake_configure_options])
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/export_project.py:410]: Why not pass the option as is?

That would make it more flexible if for some reason, I have a cmake preset that defines a bunch of Cache Variables/Environment variables
```suggestion
    if additional_cmake_configure_options:
        cmake_configure_command.extend(additional_cmake_configure_options)
```

Comment (lumberyard-employee-dm) [reply scripts/o3de/o3de/export_project.py:410]: Can we perform similar logic to what we performed with the `/` slash option for the ` -cco` option?

Comment (spham-amzn) [reply scripts/o3de/o3de/export_project.py:410]: I can't think of a clean way to represent that. Maybe have the first '/' represent the config options and any trailing one after that represents the --build options?

**Review by lumberyard-employee-dm — APPROVED** [review]
